confluent_kafka/schema_registry/config.py

```diff
@@ -0,0 +1,295 @@
+
+        _headers = self.default_headers
+
+        if body is not None:
+            body = json.dumps(body, cls=encoder)
+            _headers = {'Content-Length': str(len(body)),
+                        'Content-Type': "application/vnd.schemaregistry.v1+json"}
+
+        response = super(DefaultRestClient, self).request(method, url,
+                                                          headers=_headers,
+                                                          data=body)
+
+        try:
+            if 200 <= response.status_code <= 299:
+                return response.json(cls=decoder, **kwargs)
+            raise SchemaRegistryClientError(response.status_code,
+                                            response.json().get('message'))
+        except ValueError:
+            raise SchemaRegistryClientError(response.status_code,
+                                            response.content)
+
+
+class BasicAuthCredentialSource(object):
+    """
+    Specifies the configuration property(ies) that provide the basic
+        authentication credentials.
+    """
+    USERINFO = "USERINFO"
+    SASL_INHERIT = "SASL_INHERIT"
+    URL = "URL"
+
+
+class SchemaRegistryConfig(Config):
+    """
+
+    Keyword Args:
+        - url (str): URL for Schema Registry instance.
+
+        - schema.registry.ca.location (str, optional):
+        - schema.registry.certificate.location (str, optional):
+        - schema.registry.key.location (str, optional):
+
+        - schema.registry.basic.auth.credentials.source (str, optional):
+            Specifies the configuration property(ies) that provide the basic
+            authentication credentials.
+
+            USER_INFO: Credentials are specified via the
+            ``schema.registry.basic.auth.user.info`` config property.
+
+            SASL_INHERIT: Credentials are specified via the ``sasl.username``
+            and ``sasl.password`` configuration properties.
+
+            URL(default): Credentials are pulled from the URL if available.
+
+        - schema.registry.basic.auth.user.info (str, optional): Basic auth
+            credentials in the form {username}:{password}.
+        - schema.registry.sasl.username (str, optional): SASL username for use
+            with the PLAIN mechanism.
+        - schema.registry.sasl.password (str, optional): SASL password for use
+            with the PLAIN mechanism.
+        - schema.registry.sasl.mechanism (str, optional): SASL mechanism to
+            use for authentication. Only ``PLAIN`` is supported by
+            ``basic.auth.credentials.source`` SASL_INHERIT.
+
+    """
+    properties = {
+        URL: None,
+        CA_LOCATION: None,
+        CERTIFICATE_LOCATION: None,
+        KEY_LOCATION: None,
+        BASIC_AUTH_CREDENTIALS_SOURCE: BasicAuthCredentialSource.URL,
+        BASIC_AUTH_USER_INFO: None,
+        SASL_USERNAME: None,
+        SASL_PASSWORD: None,
+        SASL_MECHANISM: 'GSSAPI',
+        REST_CLIENT: DefaultRestClient
```

Comment:
does this need to be configurable?

also, the stuff we're making is a bit tricky, in that we have things that are configuration like that are programmatic (serdes, subject name strategies), but i think an API that explicitly separates the two is going to be better. i don't know exactly what this looks like in python.

Reply:
Actually the idea was to make it easy for me to inject mocks without hard coding it into the signature as it was previously.